doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,364 @@
+3. wait and finalize
+4. return to the user
+
+*In v2, steps 3 and 4 may be swapped.
+
+__Interpreting the Query__
+This will use standard parsing/planning, the only notable
+callout here is that we'll want `mz_sql::plan::AlterClusterOptions`
+to contain an `alter_mechanaism` field which holds an optional
+`AlterClusterMechanism` which will look something like:
+
+```rust
+enum AlterClusterMechanism {
+  Duration{
+    timeout: Duration
+    ...
+  },
+  UntilCaughtUp{
+    timeout: Duration,
+    continue_on_timeout: bool
+    ...
+  },
+}
+```
+
+__Cluster Sequencer__
+
+The following roughly outlines the new logic for the cluster sequencer when
+performing an alter cluster on a managed cluster.
+
+Because this will need to be cancelable and will wait for a significant
+duration before responding, the current `sequence_cluster_alter` process cannot
+be used. This implementation will rely on migrating `sequence_cluster_alter` to
+use the `sequence_staged` flow, allowing us to kick off a thread/stage which
+performs waits/validations.
+
+The new `sequence_alter_cluster` function will need to perform
+the current validations, used in both `sequence_alter_cluster` and
+`sequence_alter_cluster_managed_to_managed`, but will kick off a new
+`StageResult::Handle` when an `alter_mechanism` is provided. The task in this
+handle will perform waits/validations and pass off the finalization to
+a `Finalize` stage.
+
+Additionally, we'll need to perform some extra validations to ensure the
+the `alter_mechanism` and `AlterClusterPlan` are compatible. These are:
+1. validate the cluster is managed
+2. validate that the cluster is not undergoing a reconfiguration
+3. validate the cluster has no sources and now sinks
+4. the cluster cannot have a refresh schedule (v1)
+
+The `sequence_alter_cluster_managed_to_managed` method will need
+to be updated to apply the correct name/pending status for new replicas
+and only remove existing replicas when no alter_mechanism is provided.
+
+
+For V2:
+We will need to implement a message strategy that kicks the work off
+off of the `sequence_staged` flow. In this scenario, we will still create new
+replicas in the foreground, but the wait/checks will be performed in a separate
+message handle by emitting a `ClusterReconfigureCheckReady` message.
+
+The state for the reconfiguration will largely be stored in a `PendingClusterReconfiguration`
+struct.
+
+
+```rust
+struct PendingClusterReconfiguration {
+ ctx: ExecuteContext,
+ // needed to determine timeouts
+ start_time: Instant,
+ // the plan will hold the cluster_id, config, mechanism, etc...
+ plan: AlterClusterPlan,
+ // idempotency key will need to be part of the  message and will need to be
+ // validated to ensure we're looking at the correct plan, in case we
+ // miss a cancel and a new entry was put in pending_cluster_reconfiguration.
+ idempotency_key: Uuid,
+}
+```
+
+In order to cancel we'll keep a map, `pending_cluster_reconfigurations`,  of
```

Comment:
This is no longer true due to sequence_staged. This will be a new field in ConnMeta similar to its drop_sinks.